gdnative-core/src/nativescript/init/method.rs

```diff
@@ -0,0 +1,491 @@
+    iter: std::slice::Iter<'a, &'a Variant>,
+}
+
+impl<'a> Varargs<'a> {
+    /// Returns the amount of arguments left.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.iter.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns a builder for reading the next argument, that can be used to refine
+    /// the error message on failure.
+    #[inline]
+    pub fn read<T: FromVariant>(&mut self) -> ArgBuilder<'_, 'a, T> {
+        ArgBuilder {
+            args: self,
+            name: None,
+            ty: None,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Parses a structure that implements `FromVarargs` incrementally from the
+    /// remaining arguments.
+    #[inline]
+    pub fn read_many<T: FromVarargs>(&mut self) -> Result<T, Vec<ArgumentError<'a>>> {
+        T::read(self)
+    }
+
+    /// Returns the remaining arguments as a slice of `Variant`s.
+    #[inline]
+    pub fn as_slice(&self) -> &'a [&'a Variant] {
+        self.iter.as_slice()
+    }
+
+    /// Discard the rest of the arguments, and return an error if there is any.
+    ///
+    /// # Errors
+    ///
+    /// If there are any excess arguments left.
+    #[inline]
+    pub fn done(self) -> Result<(), ArgumentError<'a>> {
+        if self.is_empty() {
+            Ok(())
+        } else {
+            Err(ArgumentError::ExcessArguments {
+                rest: self.as_slice(),
+            })
+        }
+    }
+
+    /// Create a typed interface from raw pointers. This is an internal interface.
+    ///
+    /// # Safety
+    ///
+    /// `args` must point to an array of valid `godot_variant` pointers of at least `num_args` long.
+    #[doc(hidden)]
+    #[inline]
+    pub unsafe fn from_sys(num_args: libc::c_int, args: *mut *mut sys::godot_variant) -> Self {
+        let args = std::slice::from_raw_parts(args, num_args as usize);
+        let args = std::mem::transmute::<&[*mut sys::godot_variant], &[&Variant]>(args);
+        Self {
+            idx: 0,
+            iter: args.iter(),
+        }
+    }
+}
+
+impl<'a> Iterator for Varargs<'a> {
+    type Item = &'a Variant;
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().copied()
+    }
+}
+
+/// Trait for structures that can be parsed from `Varargs`.
+///
+/// This trait can be derived for structure types where each type implements `FromVariant`.
+/// The order of fields matter for this purpose:
+///
+/// ```ignore
+/// #[derive(FromVarargs)]
+/// struct MyArgs {
+///     foo: i32,
+///     bar: String,
+///     #[opt] baz: Option<Ref<Node>>,
+/// }
+/// ```
+pub trait FromVarargs: Sized {
+    fn read<'a>(args: &mut Varargs<'a>) -> Result<Self, Vec<ArgumentError<'a>>>;
+}
+
+/// Builder for providing additional argument information for error reporting.
+pub struct ArgBuilder<'r, 'a, T> {
+    args: &'r mut Varargs<'a>,
+    name: Option<&'a str>,
+    ty: Option<&'a str>,
```

Comment:
Will `name` and `ty` have the same issue? If they have different lifetimes?

Reply:
This one does turn out to be problematic since all error objects need to have the same lifetime parameter, which will have to be `'a` since `'r` is temporary. I think it might be best to store a `Cow<'r, str>` here instead, since we don't want to allocate all the parameter names when the vast majority of them are just going to be `'static`.